### PR TITLE
Solving check cv

### PR DESCRIPTION
**Pull request by ahoyosid**
We simply change fixes to read the cross_validation module instead of model_selection.
By doing  this we solve the conflicts on #698. In addition, we change the import through the files that use check_cv.

### CODE REVIEW

**Comment by ahoyosid** [comment]
@KamalakerDadi @mrahim @AlexandreAbraham @lesteve
Can someone of you review this?

File: nilearn/_utils/fixes/__init__.py [hunk 1]
```diff
@@ -10,11 +10,8 @@
     from sklearn.feature_selection import f_regression
 
 try:
-    if LooseVersion(sklearn.__version__) >= LooseVersion('0.18'):
-        from sklearn.model_selection import check_cv
-    # 0.18 > scikit-learn >= 0.16
-    else:
-        from sklearn.cross_validation import check_cv
+    #XXX this is going to be deprecated in sklearn 0.20
```
Comment (KamalakerDadi) [inline nilearn/_utils/fixes/__init__.py]: Sorry, I didn't understood your change or if I am missing. What do we do when people have scikit-learn version < 0.20 ?

**Comment by KamalakerDadi** [comment]
General comment: What is the motivation of this PR ? Is because of conflicts ?

File: nilearn/_utils/fixes/__init__.py (hunk 1 above)
Comment (ahoyosid) [inline nilearn/_utils/fixes/__init__.py]: The change is simple, not importing check_cv from model_selection, as it does not receive an X param and causes conflicts.
My bad, that comment was a personal message, I'll remove it.

File: nilearn/_utils/fixes/__init__.py [hunk 2]
```diff
@@ -10,11 +10,8 @@
     from sklearn.feature_selection import f_regression
 
 try:
-    if LooseVersion(sklearn.__version__) >= LooseVersion('0.18'):
-        from sklearn.model_selection import check_cv
     # 0.18 > scikit-learn >= 0.16
-    else:
-        from sklearn.cross_validation import check_cv
+    from sklearn.cross_validation import check_cv
```
Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/__init__.py]: I don't understand. That's the wrong change: it is importing the deprecated check_cv in 0.18. We shouldn't use the sklearn.cross_validation module in sklearn > 0.18, we should use sklearn.model_selection.

Comment (ahoyosid) [inline nilearn/_utils/fixes/__init__.py]: Ok, my bad (again).
The cross_validation module is going to be deprecated in 0.20 not 0.18.
The check_cv in model_selection does not receive X as param, as well as the CV API. Then, we have to think about this deprecation in near future --e.g. using cross validation in Nilearn examples.
For the moment we can still use the check_cv in cross_validation.

**Comment by GaelVaroquaux** [comment]
> The cross_validation module is going to be deprecated in 0.20 not 0.18.

No: it's deprecate in 0.18, it's going to be remove in 0.20. We shouldn't
be using it.

> For the moment we can still use the check_cv in cross_validation.

No, not when 0.18 is installed. We need to make an adaption layer.

**Comment by KamalakerDadi** [comment]
>We need to make an adaption layer.

I guess that is what we have in master now adaptation.

**Comment by ahoyosid** [comment]
@KamalakerDadi @GaelVaroquaux I added a backport for the check_cv and changed its use in space_net.py and group_sparse_cov.py. Now these files are using the model_selection API.

Can you check it out?

**Comment by ahoyosid** [comment]
Can you check this PR again?

ping @GaelVaroquaux  @KamalakerDadi

**Comment by bthirion** [comment]
I'm a bit surprised by the coverage decrease, but LGTM otherwise.

**Comment by KamalakerDadi** [comment]
>We need to make an adaption layer.
>Can you check this PR again?

Why don't we do like this ?

We keep the same as it is in `nilearn/_utils/fixes/__init__.py` and
make an adaptation layer in `nilearn/connectome/group_sparse_cov.py`
```python
            if LooseVersion(sklearn.__version__) >= LooseVersion('0.18'):
                cv.append(check_cv(self.cv, np.ones(subjects[k].shape[0]),
                                   classifier=False).split(subjects[k]))
            else:
                cv.append(check_cv(self.cv, subjects[k], None,
                                   classifier=False))
```
and in `nilearn/decoding/space_net.py`
```python
            if LooseVersion(sklearn.__version__) >= LooseVersion('0.18'):
                self.cv_ = list(check_cv(self.cv, y=y,
                                         classifier=self.is_classif).split(X, y))
            else:
                self.cv_ = list(check_cv(self.cv, X=X, y=y,
                                         classifier=self.is_classif))
```

In that sense, we don't need new files you have added. Sorry, if I am missing something here ?
nilearn/_utils/fixes/funcsigs.py
nilearn/_utils/fixes/sklearn_model_selection.py
nilearn/_utils/fixes/sklearn_validation.py

**Comment by KamalakerDadi** [comment]
ping @GaelVaroquaux ?

**Comment by ahoyosid** [comment]
Well, the thing is that the cross-validation objects were moved to sklearn.model_selection, and additionally they changed the API, so if we cannot use the old sklearn.cross_validation module.
I mean, you can call it as is implemented in _utils/fixes/__init__.py. However, this relies in the former API (it does not uses the cv.split(X, y)).
Then, to use the new API we have to make a wrapper, this is sklearn_model_selection. 
This module uses a signature function (funcsigs.signature, also present in sklearn.externals.funcsgs >= 0.17).
This is the reason why I had to add it.

Regarding sklearn_validation,  it is already part of master.

File: nilearn/_utils/fixes/funcsigs.py [hunk 3]
```diff
@@ -0,0 +1,814 @@
+# Copyright 2001-2013 Python Software Foundation; All Rights Reserved
```
Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/funcsigs.py]: Why is this file added?

Are you sure that it's not the result of an operation error with git?

File: nilearn/_utils/fixes/sklearn_validation.py [hunk 4]
```diff
@@ -399,3 +399,27 @@ def check_is_fitted(estimator, attributes, msg=None, all_or_any=all):
 
     if not all_or_any([hasattr(estimator, attr) for attr in attributes]):
         raise NotFittedError(msg % {'name': type(estimator).__name__})
+
+
+def indexable(*iterables):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/sklearn_validation.py]: I don't see where we are using this function. It is needed?

Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/sklearn_validation.py]: Sorry, please discard this comment. It was a mistake that I made with the diff tool.

**Comment by GaelVaroquaux** [comment]
I had a little look, and I actually really like @KamalakerDadi 's solution: it's very simple and comes with little code. I would push for adopting it.

It should of course be accompanied with the proper comments.

**Comment by ahoyosid** [comment]
Ok, so I'll change this PR

**Comment by ahoyosid** [comment]
Can you check out this PR ?
If you click the details of the continuous-integration it passes all versions. Yet, it appears as it is still processing.

Ping @KamalakerDadi @GaelVaroquaux

File: nilearn/_utils/fixes/__init__.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
         from sklearn.cross_validation import check_cv
 except ImportError:
     # scikit-learn < 0.16
-    from sklearn.cross_validation import _check_cv as check_cv
+	from sklearn.cross_validation import _check_cv as check_cv
```
Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/__init__.py]: Why this change?

File: nilearn/_utils/fixes/sklearn_validation.py (hunk 4 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/fixes/sklearn_validation.py]: Is this function used?

It seems to me that it is not any longer.

File: nilearn/connectome/group_sparse_cov.py [hunk 6]
```diff
@@ -5,6 +5,8 @@
 # Authors: Philippe Gervais
 # License: simplified BSD
 
+from distutils.version import LooseVersion
+import sklearn
```
Comment (GaelVaroquaux) [inline nilearn/connectome/group_sparse_cov.py]: This import should be put next to the other imports of sklearn.

File: nilearn/connectome/group_sparse_cov.py [hunk 7]
```diff
@@ -957,13 +966,21 @@ def fit(self, subjects, y=None):
...
+                                 n_alphas)[::-1]
 
         covs_init = itertools.repeat(None)
+
+        # Copying the cv generators to use them n_refinements times.
+        cv_list = []
+        cv_ = izip(*cv)
+        for i in range(n_refinements):
+            _, cv_ = itertools.tee(cv_)
+            cv_list.append(cv_)
```
Comment (GaelVaroquaux) [inline nilearn/connectome/group_sparse_cov.py]: Wouldn't it be a better strategy to do this copy in the loop below? Also using the tee maybe?

Comment (GaelVaroquaux) [inline nilearn/connectome/group_sparse_cov.py]: Just this comment to address before the merge. To be clear, what I have in mind is something like:

</pre>
n_refinements = 3

cv = (i**2 for i in range(5))

import itertools

# Use itertools.tee to duplicate multiple times the cv iterator
for this_cv in itertools.tee(cv, n_refinements):
    print(list(this_cv))
</pre>

And then you can use the cv object just as before.

**Comment by GaelVaroquaux** [comment]
I want to merge this PR, but there is one last comment to address. I gave more details to make sure that you understand it.

**Comment by ahoyosid** [comment]
Ok, I've made the changes. 
Thank you for the insight !

**Comment by GaelVaroquaux** [comment]
Merging! Thanks!
